### ISSUE TITLE
How to generate registerFactoryAsync and getIt.getAsync<T>()

### ISSUE THREAD

**Issue by RakaAlrian** [Milad-Akarie/injectable#186]
How to generate` registerFactoryAsync<T>()` instead of `registerFactory<T>()` and use `getIt.getAsync<T>()` to fill parameter instead of `getIt<T>()` ?

**Comment by Milad-Akarie** [comment]
Hey @RakaAlrian Please refer to the docs here 
https://pub.dev/packages/injectable#registering-asynchronous-injectables

**Comment by RakaAlrian** [comment]
@Milad-Akarie thank you for your fast response.
But I still can't figure how to implement the docs example to my project, if you have time please kindly look to these files.

retrofit.dart
```dart
class Retrofit {
  final DeviceInfo deviceInfo;
  PackageInfo packageInfo;
  String baseUrl;
  int timeOut;
  final EmailDao emailDao;

  Retrofit({
    @required this.deviceInfo,
    @required this.packageInfo,
    this.baseUrl = ConstData.baseUrl,
    this.timeOut = 10000,
    @required this.emailDao,
  });

  Future<Dio> getDioInstance() async => Dio(await baseOptions());

  Future<BaseOptions> baseOptions() async {
    //some code
    return BaseOptions(
      //some code
    );
  }
}
```
sign_in_service.dart
```dart
part 'sign_in_service.g.dart';

@RestApi()
abstract class SignInService {

  factory SignInService(Dio dio) = _SignInService;

  static const _defaultPath = "/users/auth";

  @POST("_defaultPath")
  Future<BaseResponse<UserModel>> getUserData();

}
```
Here is how I managed to inject them manually using get_it
app_injection.dart
```dart
import 'app_injection.config.dart';

final inject = GetIt.instance;

@InjectableInit()
Future<void> configureDependencies() async {
  await $initGetIt(inject);

  // RETROFIT
  inject.registerFactoryAsync<Dio>(() => Retrofit(
      packageInfo: inject<PackageInfo>(),
      deviceInfo: inject<DeviceInfo>(),
      emailDao: inject<EmailDao>()
  ).getDioInstance());
  // SERVICES
  inject.registerFactoryAsync<SignInService>(() async => SignInService(await inject.getAsync<Dio>()));
}
```
The stuck point is in this line `inject.registerFactoryAsync<SignInService>(() async => SignInService(await inject.getAsync<Dio>()));`, I don't know how to generate `getAsync<Dio>` using injectable. I don't think I can do much with sign_in_service.dart file since it's a template from [retrofit](https://pub.dev/packages/retrofit) package.

**Comment by Milad-Akarie** [comment]
@RakaAlrian Your retrofit service requires an instance of Dio so I guess you have to preResolve It's value

try this
```dart
// make your retrofit class Injectable
@injectable
class Retrofit{
 ..... 
}
```

create a register module 
```dart
@module
abstract class DioModule {
// the preResolve annotation will have Dio register as an instance not an asynchronous dependency. 
  @preResolve
  @LazySingleton
  Future<Dio>  getDioInstance (Retrofit retrofit) => retrofit.getDioInstance(); // injectable will inject Retrofit for you
}
```

don't forget to await for `configureDependencies() => $initGetIt(inject);`

**Comment by RakaAlrian** [comment]
@Milad-Akarie sorry for slow response. I've tried your suggestion and that's working perfectly, I didn't know about put parameter inside function module like that before. I also tried to use `@factoryMethod` inside my SignInService and that's working too, turns out `@factoryMethod` can work for synchronous too.

Here is my sign_in_service.dart now.

```dart
part 'sign_in_service.g.dart';

@RestApi()
abstract class SignInService {

  factory SignInService(Dio dio) = _SignInService;

  static const _defaultPath = "/users/auth";

  @POST("_defaultPath")
  Future<BaseResponse<UserModel>> getUserData();
}
```
In the end I don't have to use get_it manually, thank you for your response and of course your amazing package😁
